source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -61,9 +98,22 @@ class CacheFilter : public Http::PassThroughFilter,
   // https://httpwg.org/specs/rfc7234.html#response.cacheability
   bool request_allows_inserts_ = false;
 
+  // True if the CacheFilter is injected validation headers and should check for 304 responses
+  bool validating_cache_entry_ = false;
+
+  // True if CacheFilter::encodeHeaders & CacheFilter::encodeData should be skipped
+  bool skip_encoding_ = false;
+
   // Used for coordinating between decodeHeaders and onHeaders.
   enum class GetHeadersState { Initial, FinishedGetHeadersCall, GetHeadersResultUnusable };
-  GetHeadersState state_ = GetHeadersState::Initial;
+  GetHeadersState get_headers_state_ = GetHeadersState::Initial;
+
+  // Used for coordinating between encodeHeaders/encodeData and onBody/onTrailers
+  enum class EncodeCachedResponseState { Initial, FinishedEncoding, IterationStopped };
+  EncodeCachedResponseState encode_cached_response_state_ = EncodeCachedResponseState::Initial;
+
+  enum class FilterState { Initial, Decoding, Encoding };
```

Comment:
`GetHeadersState` & `EncodeCachedResponseState` operate in completely difference scenarios. However, it will indeed be a lot nicer and easier to understand if one state is enough. I will look into it.